pkg/rosa/oidcconfigs/oidcconfigs.go

```diff
@@ -0,0 +1,217 @@
+package oidcconfigs
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"gopkg.in/square/go-jose.v2"
+
+	"github.com/openshift-online/ocm-common/pkg/utils"
+)
+
+const (
+	defaultLengthRandomLabel = 4
+
+	prefixForPrivateKeySecret     = "rosa-private-key"
+	defaultPrefixForConfiguration = "oidc"
+)
+
+type OidcConfigInput struct {
+	BucketName           string
+	IssuerUrl            string
+	PrivateKey           []byte
+	PrivateKeyFilename   string
+	DiscoveryDocument    string
+	Jwks                 []byte
+	PrivateKeySecretName string
+}
+
+func BuildOidcConfigInput(userPrefix, region string) (OidcConfigInput, error) {
+	bucketName, err := GenerateBucketName(userPrefix)
+	if err != nil {
+		return OidcConfigInput{}, fmt.Errorf("There was a problem generating bucket name: %s", err)
+	}
+
+	privateKeySecretName := fmt.Sprintf("%s-%s", prefixForPrivateKeySecret, bucketName)
+	bucketUrl := fmt.Sprintf("https://%s.s3.%s.amazonaws.com", bucketName, region)
+	privateKey, publicKey, err := CreateKeyPair()
+	if err != nil {
+		return OidcConfigInput{}, fmt.Errorf("There was a problem generating key pair: %s", err)
+	}
+	privateKeyFilename := fmt.Sprintf("%s.key", privateKeySecretName)
+	discoveryDocument := GenerateDiscoveryDocument(bucketUrl)
+	jwks, err := BuildJSONWebKeySet(publicKey)
+	if err != nil {
+		return OidcConfigInput{}, fmt.Errorf("There was a problem generating JSON Web Key Set: %s", err)
+	}
+	return OidcConfigInput{
+		BucketName:           bucketName,
+		IssuerUrl:            bucketUrl,
+		PrivateKey:           privateKey,
+		PrivateKeyFilename:   privateKeyFilename,
+		DiscoveryDocument:    discoveryDocument,
+		Jwks:                 jwks,
+		PrivateKeySecretName: privateKeySecretName,
+	}, nil
+}
+
+func GenerateBucketName(userPrefix string) (string, error) {
+	randomLabel := utils.RandomLabel(defaultLengthRandomLabel)
+	bucketName := fmt.Sprintf("%s-%s", defaultPrefixForConfiguration, randomLabel)
+	if userPrefix != "" {
+		bucketName = fmt.Sprintf("%s-%s", userPrefix, bucketName)
+	}
+	if !IsValidBucketName(bucketName) {
+		return "", fmt.Errorf("The bucket name '%s' is not valid", bucketName)
+	}
+
+	return bucketName, nil
+}
+
+const (
+	bucketNameRegex = "^[a-z][a-z0-9\\-]+[a-z0-9]$"
+)
+
+func IsValidBucketName(bucketName string) bool {
+	if bucketName[0] == '.' || bucketName[len(bucketName)-1] == '.' {
+		return false
+	}
+	if strings.HasPrefix(bucketName, "xn--") {
+		return false
+	}
+	if strings.HasSuffix(bucketName, "-s3alias") {
+		return false
+	}
+	if match, _ := regexp.MatchString("\\.\\.", bucketName); match {
+		return false
+	}
+	// We don't support buckets with '.' in them
+	match, _ := regexp.MatchString(bucketNameRegex, bucketName)
+	return match
+}
+
+func CreateKeyPair() ([]byte, []byte, error) {
```

Comment:
Would it not be better to return the public key in DER format rather than marshalising it twice and decoding it in `BuildJSONWebKeySet`

Reply:
Can't say it would be better without investigating, right now it's just how the callers are expecting it